### ISSUE TITLE
Unable to Change Hosts for Vanity URLS - 5.3.8 and possibily 21.06

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#21772]
Copied from: #21373 

Found when working on Support Ticket: [105909](https://dotcms.zendesk.com/agent/tickets/105909)
Reproducible as of dotCMS 5.3.8.6 up to latest master (21.11)

Vanity URLs can work for either a specific Site or all Sites in a dotCMS instance. If you change where a given Vanity URL is supposed to work on, it should work as expected. However, when doing so, changes are not getting picked up and the Vanity URL still points to the previous  value -- either a specific Site, or all Sites.

Flushing the Cache Regions and/or doing a full reindex is not fixing the problem.

### NOTE TO DEVS #2: At least in dotCMS 5.3.8.6, adding the `Site or Folder` field does not solve the problem. The back-end is ignoring any change you make to it and the new Site you select is not persisted.

**To Reproduce**
1. Create a copy of the `demo.dotcms.com` Site named `second.dotcms.com`.
2. Select the `demo.dotcms.com` Site, and create a Vanity URL with the following information. **It's very important the the Vanity points to `All sites` first**:
2.1. Title: `My test Vanity`
2.2. Site: `All sites`
2.3. Uri: `/my-test`
2.4. Action: `301 - Redirect`
2.5. Forward To: `/index`
3. The Vanity URL works for `demo.dotcms.com`:
```
$ curl -sk --head -H "Host: demo.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:42 GMT
Connection: close
```
4. And for `second.dotcms.com` as well, which is expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:58 GMT
Connection: close
```
5. Now, change the Vanity URL to point to `demo.dotcms.com` **ONLY**, and try both CURL commands again.
6. Both of them are returning Status Code 301, which is **NOT** correct as the one sent to `second.dotcms.com` should have failed with a 404.
7. If you re-create the Vanity URL and point it to the `demo.dotcms.com` Site since the beginning, the CURL command to `demo.dotcms.com` works again, and the one to `second.dotcms.com` fails as expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 404 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Set-Cookie: JSESSIONID=D6BA571A1954E4ABD0E9A5951245AEFE; Path=/; Secure; HttpOnly; SameSite=Strict
Content-Type: text/html;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:45:09 GMT
Connection: close
```

**Expected behavior**
Changing the **Site** field must update the information of the Vanity URL correctly.

**Additional context**
The only workaround is to re-create the Vanity URL from scratch. However, this is not a feasible solution for the customer. We need to look at the possibility of creating a SQL query that could update the incorrect and fix this automatically.

In the customer instance, running 5.3.8.6, I was able to confirm that the "more technical" workaround for fixing a Vanity URL that is not working is the following:

- Look for the Vanity URL in the `identifier` table, and update the `host_inode` column to point to either the right site or `SYSTEM_HOST` if you need it to apply to all Sites.
- Flush the region called `Vanity URL Cache`.
- Test the Vanity URL once more. You can use CURL, like this:
```
curl -sk --head -H "Host: www.yoursite.com" https://your-dotcms-instance.com/your-vanity-url
```

**Comment by wezell** [comment]
We need to fix this against the `5.3.8.9` branch

**Comment by ian-b-cooper** [comment]
Quality Control Tested by Non-Dev. Fix confirmed

**Comment by swicken-dotcms** [comment]
Reverted in 21.06.7 as it created other issues with Vanity URLs

**Comment by victoralfaro-dotcms** [comment]
22.04
![image](https://user-images.githubusercontent.com/64026161/159590602-f31d9a88-d5bd-4736-b63e-1da1c989d09b.png)
All sites:
```
curl -sk --head -H "Host: demo.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|95e7de1ba5
X-DOT-VanityUrl: 52ca89ad38a16114443757bdc74136d4
Location: /index
Transfer-Encoding: chunked
Date: Tue, 22 Mar 2022 23:03:09 GMT

curl -sk --head -H "Host: second.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|95e7de1ba5
X-DOT-VanityUrl: 52ca89ad38a16114443757bdc74136d4
Location: /index
Transfer-Encoding: chunked
Date: Tue, 22 Mar 2022 23:03:46 GMT
```
demo.dotcms.com
```
curl -sk --head -H "Host: demo.dotcms.com"  http://localhost:8080/my-test              ✔  17:03:46 
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|95e7de1ba5
X-DOT-VanityUrl: 52ca89ad38a16114443757bdc74136d4
Location: /index
Transfer-Encoding: chunked
Date: Tue, 22 Mar 2022 23:05:36 GMT

curl -sk --head -H "Host: second.dotcms.com"  http://localhost:8080/my-test                  ✔  17:05:36 
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|95e7de1ba5
X-DOT-VanityUrl: 52ca89ad38a16114443757bdc74136d4
Location: /index
Transfer-Encoding: chunked
Date: Tue, 22 Mar 2022 23:05:41 GMT
```

**Comment by victoralfaro-dotcms** [comment]
21.06.7
All sites
```
curl -sk --head -H "Host: demo.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 4847338a-7000-4e04-baba-5a7ae9214e27
Location: /index
Transfer-Encoding: chunked
Date: Wed, 23 Mar 2022 01:46:00 GMT

curl -sk --head -H "Host: second.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 4847338a-7000-4e04-baba-5a7ae9214e27
Location: /index
Transfer-Encoding: chunked
Date: Wed, 23 Mar 2022 01:46:09 GMT
```
demo.dotcms.com
```
```

**Comment by victoralfaro-dotcms** [comment]
Testing against `22.04`
```
curl -sk --head -H "Host: demo.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|816cfb79bd
X-DOT-VanityUrl: ac2669d3486e0418df8441899f929568
Location: /index
Transfer-Encoding: chunked
Date: Thu, 31 Mar 2022 14:27:23 GMT

curl -sk --head -H "Host: second.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|816cfb79bd
X-DOT-VanityUrl: ac2669d3486e0418df8441899f929568
Location: /index
Transfer-Encoding: chunked
Date: Thu, 31 Mar 2022 14:27:38 GMT
```
After changing vanity url's site to `demo.dotcms.com`
```
curl -sk --head -H "Host: demo.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 301
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|816cfb79bd
X-DOT-VanityUrl: ac2669d3486e0418df8441899f929568
Location: /index
Transfer-Encoding: chunked
Date: Thu, 31 Mar 2022 14:27:59 GMT

curl -sk --head -H "Host: second.dotcms.com"  http://localhost:8080/my-test
HTTP/1.1 404
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
x-dot-server: unknown|816cfb79bd
Set-Cookie: JSESSIONID=BA31F4F1E58AC021500EA0FCC8231E74; Path=/; Secure; HttpOnly; SameSite=Strict
vary: accept-encoding
Content-Type: text/html;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Thu, 31 Mar 2022 14:28:02 GMT
```

**Comment by bryanboza** [comment]
Fixed, tested following the provided steps and now this is working as expected, also tested doing multiple changes and works fine.

**Comment by bryanboza** [comment]
This change was broken by a recent change  #21889

**Comment by dsilvam** [comment]
Passed Internal QA: working as expected after following the steps in description. 
After changing the Site to `demo.dotcms.com` the second curl gives 404

**Comment by bryanboza** [comment]
Fixed, tested on release-22.04 // Docker // FF
